Synchronization error after upgrading server to CouchDB 2.3

Please [read these guidelines](http://ibm.biz/cdt-issue-guide) before opening an issue.

<!-- Issues will be CLOSED IMMEDIATELY if the following template is not completed. -->

## Bug Description
After upgrading server to CouchDB 2.3 the synchronization is failling with the following exception:

```
1560 15126 W RetriableTask: Received an exception during response stream processing. A retry will be attempted.
01-03 02:31:25.336 31560 15126 W RetriableTask: java.lang.RuntimeException: com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `long` from String "0-g1AAAABXeJzLYWBgYMpgTmEQTM4vTc5ISXIwNDLXMwBCwxyQVCJDUv3___-zEhnwKMpjAZIMDUDqP0RtFgDZJhhE": not a valid Long value
01-03 02:31:25.336 31560 15126 W RetriableTask:  at [Source: (InputStreamReader); line: 1, column: 39] (through reference chain: com.cloudant.sync.internal.mazha.CouchDbInfo["purge_seq"])
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.cloudant.sync.internal.util.JSONUtils.fromJson(JSONUtils.java:209)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.cloudant.sync.internal.mazha.CouchClient$TypeInputStreamProcessor.processStream(CouchClient.java:793)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.cloudant.sync.internal.mazha.CouchClient.executeWithRetry(CouchClient.java:215)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.cloudant.sync.internal.mazha.CouchClient.executeWithRetry(CouchClient.java:259)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.cloudant.sync.internal.mazha.CouchClient.executeToJsonObjectWithRetry(CouchClient.java:244)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.cloudant.sync.internal.mazha.CouchClient.executeToJsonObjectWithRetry(CouchClient.java:249)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.cloudant.sync.internal.mazha.CouchClient.getDbInfo(CouchClient.java:279)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.cloudant.sync.internal.replication.CouchClientWrapper.exists(CouchClientWrapper.java:76)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.cloudant.sync.internal.replication.PushStrategy.replicate(PushStrategy.java:201)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.cloudant.sync.internal.replication.PushStrategy.run(PushStrategy.java:165)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at java.lang.Thread.run(Thread.java:764)
01-03 02:31:25.336 31560 15126 W RetriableTask: Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `long` from String "0-g1AAAABXeJzLYWBgYMpgTmEQTM4vTc5ISXIwNDLXMwBCwxyQVCJDUv3___-zEhnwKMpjAZIMDUDqP0RtFgDZJhhE": not a valid Long value
01-03 02:31:25.336 31560 15126 W RetriableTask:  at [Source: (InputStreamReader); line: 1, column: 39] (through reference chain: com.cloudant.sync.internal.mazha.CouchDbInfo["purge_seq"])
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.fasterxml.jackson.databind.DeserializationContext.weirdStringException(DeserializationContext.java:1548)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.fasterxml.jackson.databind.DeserializationContext.handleWeirdStringValue(DeserializationContext.java:910)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$LongDeserializer._parseLong(NumberDeserializers.java:584)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$LongDeserializer.deserialize(NumberDeserializers.java:557)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$LongDeserializer.deserialize(NumberDeserializers.java:535)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:136)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:287)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4001)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3044)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	at com.cloudant.sync.internal.util.JSONUtils.fromJson(JSONUtils.java:207)
01-03 02:31:25.336 31560 15126 W RetriableTask: 	... 10 more
```

CouchDB 2.3 has changed the purge_seq attribute from long to string which is a breaking change for Cloudant Sync.

Te problem appears when the repplication process checks the existencie of the remote database by requesting the db informaton. 

### 1. Steps to reproduce and the simplest code sample possible to demonstrate the issue
Perform a replication with a CouchDB 2.3 server
<!--
Outline the steps you take to make the problem happen.
Provide the simplest code sample you can, in context, that reproduces the issue.
-->

### 2. What you expected to happen
Synchronize without exceptions
### 3. What actually happened
The shyncronization is stopped

## Environment details
CouchDB 2.3 on the server
<!--
- Version(s) that are affected by this issue.
    > 2.0.2
- Android or Java version (including vendor and platform)
    > Android 5.1, API level 22, Linux/armv7l
    > Java 1.8.0_151 OpenJDK macOS
- sync-android artifact you are using:
    - [X] cloudant-sync-datastore-android
    - [ ] cloudant-sync-datastore-android-encryption
    - [ ] cloudant-sync-datastore-javase
-->

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -2,7 +2,7 @@
 - [NEW] `Database` methods `read`, `contains`, `create`, and `delete` now accept local
   (non-replicating documents). These documents must have their document ID prefixed with `_local/`
   and must have their revision ID set to `null` (where applicable).
-- [FIXED] Exception parsing remote database information when replicating with CouchDB 2.3. 
+- [FIXED] Fixed purge_seq to return a String object when replicating with CouchDB 2.3 databases. 
      
 # 2.3.0 (2018-08-14)
 - [NEW] Added API for specifying a list of document IDs in the filtered pull replicator.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -2,7 +2,8 @@
 - [NEW] `Database` methods `read`, `contains`, `create`, and `delete` now accept local
   (non-replicating documents). These documents must have their document ID prefixed with `_local/`
   and must have their revision ID set to `null` (where applicable).
-  
+- [FIXED] Exception parsing remote database information when replicating with CouchDB 2.3. 
+     
 # 2.3.0 (2018-08-14)
 - [NEW] Added API for specifying a list of document IDs in the filtered pull replicator.
 - [IMPROVED] Forced a TLS1.2 `SSLSocketFactory` where possible on Android API versions < 20 (it is

--- a/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/mazha/CouchDbInfo.java
+++ b/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/mazha/CouchDbInfo.java
@@ -40,7 +40,7 @@ public class CouchDbInfo {
 	private Object updateSeq;
 
 	@JsonProperty("purge_seq")
-	private long purgeSeq;
+	private String purgeSeq;
 
 	@JsonProperty("compact_running")
 	private boolean compactRunning;
@@ -73,7 +73,7 @@ public class CouchDbInfo {
 		return updateSeq;
 	}
 
-	public long getPurgeSeq() {
+	public String getPurgeSeq() {
 		return purgeSeq;
 	}
 

--- a/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/util/JSONUtilsTest.java
+++ b/cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/util/JSONUtilsTest.java
@@ -16,17 +16,24 @@
 
 package com.cloudant.sync.util;
 
-import com.cloudant.sync.internal.util.JSONUtils;
+import com.cloudant.sync.internal.mazha.CouchDbInfo;
 import com.cloudant.sync.internal.replication.Foo;
-import org.junit.Assert;
+import com.cloudant.sync.internal.util.JSONUtils;
+import com.fasterxml.jackson.core.type.TypeReference;
+
 import org.apache.commons.io.FileUtils;
+import static org.hamcrest.CoreMatchers.is;
+import org.junit.Assert;
 import org.junit.Test;
 
+import java.io.File;
+import java.io.FileReader;
 import java.io.IOException;
 import java.util.Arrays;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.regex.Matcher;
 
 public class JSONUtilsTest {
 
@@ -147,6 +154,16 @@ public class JSONUtilsTest {
         Map<String, Object> breakfast = (Map)activities.get(1);
         Assert.assertEquals(4, breakfast.size());
         Assert.assertEquals(Integer.valueOf(40), breakfast.get("Duration"));
+    }
+
+    @Test
+    public void deserialize_CouchDBInfo_purgeSeq_from_long() throws Exception {
+        File dbinfoData = TestUtils.loadFixture("fixture/dbinfo-pre23.json");
+        CouchDbInfo deserialized = JSONUtils.fromJson(new FileReader(dbinfoData), new
+                TypeReference<CouchDbInfo>() {
+        });
+        Long purgeSeq = Long.valueOf(deserialized.getPurgeSeq());
+        Assert.assertThat(purgeSeq, is(0l));
     }
 
     private byte[] readJsonDataFromFile(String filename) throws IOException {

--- a/fixture/dbinfo-pre23.json
+++ b/fixture/dbinfo-pre23.json
@@ -1,0 +1,20 @@
+{
+  "db_name": "test",
+  "update_seq": "4-g1AAAAFTeJzLYWBg4MhgTmEQTM4vTc5ISXLIyU9OzMnILy7JAUoxJTIkyf___z8rkQGPoiQFIJlkD1bHiE-dA0hdPFgdEz51CSB19QTtzWMBkgwNQAqodD4xahdA1O4nRu0BiNr7-P0EUfsAohbk3iwA10dvMg",
+  "sizes": {
+    "file": 161030,
+    "external": 112978,
+    "active": 113762
+  },
+  "purge_seq": 0,
+  "other": {
+    "data_size": 112978
+  },
+  "doc_del_count": 0,
+  "doc_count": 3,
+  "disk_size": 161030,
+  "disk_format_version": 6,
+  "data_size": 113762,
+  "compact_running": false,
+  "instance_start_time": "0"
+}

